### PR TITLE
DEV2-1533 report suggestion about to shown

### PR DESCRIPTION
**Pull request by dimacodota**
DEV2-1533 report suggestion about to shown

### CODE REVIEW

File: src/aboutToRenderDecorator.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import TabnineInlineCompletionItem from "./inlineSuggestions/tabnineInlineCompletionItem";
+import suggestionShown from "./binary/requests/suggestionShown";
+import CompletionOrigin from "./CompletionOrigin";
+
+type DecoratedFunction<R, T1 extends vscode.TextDocument, T2, T3, T4> = (
+  document: T1,
+  ...args: [T2, T3, T4]
+) => Promise<R>;
+
+export default function aboutToRenderDecorator<
```
Comment (dimacodota) [inline src/aboutToRenderDecorator.ts:11]: this is the main functionality

File: src/inlineSuggestions/documentChangesTracker/DocumentTextChangeContent.ts [hunk 2]
```diff
@@ -50,6 +50,7 @@ export default class DocumentTextChangeContent {
       !!this.contentChange &&
       // in case of /t the rangeLength will be 1, in case of spaces the rangeLength will be tabsize
       this.contentChange.rangeLength > 0 &&
+      this.contentChange.range.isSingleLine &&
```
Comment (dimacodota) [inline src/inlineSuggestions/documentChangesTracker/DocumentTextChangeContent.ts:53]: this is an unrelated fix

Comment (dimacodota) [reply src/inlineSuggestions/documentChangesTracker/DocumentTextChangeContent.ts:53]: ill add a test for it in a different pr

File: src/binary/requests/suggestionShown.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  };
+}
+
+function suggestionShown(
+  request: SuggestionShown
+): Promise<unknown | undefined | null> {
+  return tabNineProcess.request<unknown, SuggestionShown>(request);
+}
+
+export default suggestionShown;
```
Comment (yonip23) [inline src/binary/requests/suggestionShown.ts:19]: why on a separate statement?

File: src/aboutToRenderDecorator.ts (hunk 1 above)
Comment (yonip23) [inline src/aboutToRenderDecorator.ts:11]: why like this? why not inline this functionality inside `provideInlineCompletionItems`?
this is much less readable

File: src/test/suite/completion.test.ts [hunk 4]
```diff
@@ -392,6 +393,22 @@ describe("Should do completion", () => {
       )
     ).twice();
   });
+  it.only("should report suggestion about to shown after debounce", async () => {
```
Comment (yonip23) [inline src/test/suite/completion.test.ts]: remove `.only`

File: src/test/suite/completion.test.ts [hunk 5]
```diff
@@ -392,6 +393,22 @@ describe("Should do completion", () => {
...
+    mockGetDebounceConfig(SHORT_DEBOUNCE_VALUE);
+    mockAutocomplete(requestResponseItems, anAutocompleteResponse("d", "data"));
+    await openADocWith("const ", "text");
+    await emulationUserInteraction();
+
+    await vscode.commands.executeCommand("type", {
+      text: "d",
+    });
+    await sleep(WAIT_LONGER_THAT_DEBOUNCE);
+    await emulationUserInteraction();
```
Comment (yonip23) [inline src/test/suite/completion.test.ts:406]: this function sleeps as well, why sleep twice?

File: src/test/suite/completion.test.ts [hunk 6]
```diff
@@ -392,6 +393,22 @@ describe("Should do completion", () => {
       )
     ).twice();
   });
+  it("should report suggestion about to shown after debounce", async () => {
```
Comment (yonip23) [inline src/test/suite/completion.test.ts:396]: please add a test that the event is not fired if the completion has been canceled (i.e only send events on completions that are actually about to be shown and not canceled)

File: src/test/suite/completion.test.ts [hunk 7]
```diff
@@ -44,13 +44,15 @@ import {
   setupForCompletionsTests,
   mockGetDebounceConfig,
 } from "./completion.driver";
+import { SuggestionShownRequestMatcher } from "./utils/SuggestionShownRequestMatcher";
 
 describe("Should do completion", () => {
   const SPACES_INDENTATION = "    ";
   const TAB_INDENTATION = "\t";
-  const WAIT_LONGER_THAT_DEBOUNCE = 250;
+  const WAIT_LONGER_THAN_SHORT_DEBOUNCE = 350;
```
Comment (yonip23) [inline src/test/suite/completion.test.ts:52]: remove `WAIT` from const name, its not a command, its a const (same for line 55)

File: src/provideInlineCompletionItems.ts [hunk 8]
```diff
@@ -31,10 +32,17 @@ export default async function provideInlineCompletionItems(
...
+        document,
+        completionInfo,
+        position
+      );
+      reportSuggestionShown(document, result);
+      return result;
     }
 
     const completions = await debounceCompletions(document, position, token);
+    reportSuggestionShown(document, completions);
```
Comment (yonip23) [inline src/provideInlineCompletionItems.ts:45]: are we sure the suggestion is indeed shown at this point (same for line 41)? I think vscode has its own debounce+cancel mechanism... this function just registers the plugin's suggestions to vscode - they decided later whether or not to present them, no?

**Review by yonip23 — APPROVED** [review]
lgtm - see my question in the comment

File: src/provideInlineCompletionItems.ts (hunk 8 above)
Comment (dimacodota) [reply src/provideInlineCompletionItems.ts:45]: yes, you're right, it's the closest to rendering we have right now (the shown callback will be released by the vscode team), but from what I understand from debugging their API their debounce is before this callback is executed. so, for now, it's as good as it gets
